hydro_deploy/core/src/hydroflow_crate/build.rs

```diff
@@ -1,99 +1,114 @@
-use std::collections::HashMap;
 use std::error::Error;
 use std::fmt::Display;
 use std::io::BufRead;
 use std::path::{Path, PathBuf};
 use std::process::{Command, Stdio};
-use std::sync::{Arc, Mutex};
+use std::sync::OnceLock;
 
 use cargo_metadata::diagnostic::Diagnostic;
+use memo_map::MemoMap;
 use nanoid::nanoid;
-use once_cell::sync::Lazy;
 use tokio::sync::OnceCell;
 
 use crate::progress::ProgressTracker;
 use crate::HostTargetType;
 
-#[derive(PartialEq, Eq, Hash)]
-struct CacheKey {
+/// Build parameters for [`build_crate_memoized`].
+#[derive(PartialEq, Eq, Hash, Clone)]
+pub struct BuildParams {
+    /// The working directory for the build, where the `cargo build` command will be run. Crate root.
+    /// [`Self::new`] canonicalizes this path.
     src: PathBuf,
+    /// `--bin` binary name parameter.
     bin: Option<String>,
+    /// `--example` parameter.
     example: Option<String>,
+    /// `--profile` parameter.
     profile: Option<String>,
+    /// `--target <linux>` if cross-compiling for linux ([`HostTargetType::Linux`]).
     target_type: HostTargetType,
+    /// `--features` flags, will be comma-delimited.
     features: Option<Vec<String>>,
 }
+impl BuildParams {
+    /// Creates a new `BuildParams` and canonicalizes the `src` path.
+    pub fn new(
+        src: impl AsRef<Path>,
+        bin: Option<String>,
+        example: Option<String>,
+        profile: Option<String>,
+        target_type: HostTargetType,
+        features: Option<Vec<String>>,
+    ) -> Self {
+        // `fs::canonicalize` prepends windows paths with the `r"\\?\"`
+        // https://stackoverflow.com/questions/21194530/what-does-mean-when-prepended-to-a-file-path
+        // However, this breaks the `include!(concat!(env!("OUT_DIR"), "/my/forward/slash/path.rs"))`
+        // Rust codegen pattern on windows. To help mitigate this happening in third party crates, we
+        // instead use `dunce::canonicalize` which is the same as `fs::canonicalize` but avoids the
+        // `\\?\` prefix when possible.
+        let src = dunce::canonicalize(src).expect("Failed to canonicalize path for build.");
+
+        BuildParams {
+            src,
+            bin,
+            example,
+            profile,
+            target_type,
+            features,
+        }
+    }
+}
 
-pub type BuiltCrate = Arc<(String, Vec<u8>, PathBuf)>;
+/// Information about a built crate. See [`build_crate`].
+pub struct BuildOutput {
+    /// A unique but meaningless id.
+    pub unique_id: String,
```

Comment:
Eventually we should probably use a hash; nanoid is a pretty hacky solution since we could even have collisions.